Add long-term statistics support for homematic sensors

chriss158 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds long-term statistics support for homematic component sensors.

In addition, actors like **HmIP-BSM** with a current sensor now use the device class `DEVICE_CLASS_CURRENT` instead of `DEVICE_CLASS_POWER`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chriss158 committed:
Add long-term statistics support for homematic

danielperna84 reviewed (approved):
No idea if this is in accordance with the Home Assistant standards, but technically LGTM. 👍

rytilahti commented:
Looks like this adds `ENERGY_COUNTER`, doc update needed?

chriss158 committed:
Merge branch 'dev' into homematic-long-term-statistics

chriss158 committed:
Refactor cast list to SensorEntityDescription dict

Additional:
- Gas power, gas energy counter, air pressure and voltage uses long-term-statistics
- Gas power, gas energy counter uses device class gas
- Voltage uses device class voltage
- air pressure uses device class pressure

chriss158 committed:
Merge branch 'dev' into homematic-long-term-statistics

chriss158 committed:
Refactor expensive loop to separate dictionarys

chriss158 committed:
Merge branch 'dev' into homematic-long-term-statistics

itn3rd77 commented:
Is anything blocking this from getting merged? Would bei really nice to get this with the next release. Thx!

pvizeli commented on homeassistant/components/homematic/sensor.py:
```diff
@@ -45,54 +58,163 @@
 
-HM_ICON_HA_CAST = {"WIND_SPEED": "mdi:weather-windy", "BRIGHTNESS": "mdi:invert-colors"}
+
+def get_sensor_type_desc_attr(sensor_type, attr, default=None):
```
Lets set the entity description property on the init and remove that property

pvizeli replied:
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity.py

chriss158 replied:
Makes sense. Thanks 👍

chriss158 committed:
Merge remote-tracking branch 'origin/dev' into homematic-long-term-statistics

chriss158 committed:
Merge remote-tracking branch 'upstream/dev' into homematic-long-term-statistics

chriss158 committed:
Merge remote-tracking branch 'upstream/dev' into homematic-long-term-statistics

chriss158 committed:
Use entity description property + fix humidity sensor

pvizeli commented on homeassistant/components/homematic/sensor.py:
```diff
@@ -111,6 +245,13 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        """Initialize HMSensor."""
+        super().__init__(conf)
+        self.entity_description = SENSOR_DESCRIPTIONS.get(self._state)
+        if self.entity_description is None:
```
Let's not make a fallback. At one point we have to fix that to having for all sensors the correct description object. So if somethings pop-up, we can fix it

chriss158 replied:
Without the default description it will crash, and then the sensor does not work. Maybe log an error and still do the copy? 
![grafik](https://user-images.githubusercontent.com/12832850/139723836-855d3dcd-3464-4e9e-835a-a0a2451fbd1d.png)
I would open another PR to maintain those missing sensor descriptions. Or what do you think?

pvizeli replied:
Okay but let's do that on the setup platform and update the base HMDevice `__init__(conf: Dict[str, str], description: EntityDescription | None = None)` So we don't need overload the init on each platform (future)

chriss158 replied:
Done. Thanks for review 👍

pvizeli reviewed (commented):
Looks almost good 👍

chriss158 committed:
Log missing sensor descriptions

chriss158 committed:
Merge remote-tracking branch 'upstream/dev' into homematic-long-term-statistics

bseibold commented on homeassistant/components/homematic/sensor.py:
```diff
@@ -45,54 +59,174 @@
+        key="AVERAGE_ILLUMINATION",
+        native_unit_of_measurement=LIGHT_LUX,
+        device_class=DEVICE_CLASS_ILLUMINANCE,
+        state_class=STATE_CLASS_TOTAL,
```
Are you sure `STATE_CLASS_TOTAL` is correct for this one and the next two? Shouldn't it be `STATE_CLASS_MEASUREMENT`?

chriss158 replied:
Hmm, I think you are right. I read the documentation again and it says:

> https://developers.home-assistant.io/docs/core/entity/sensor/#available-state-classes
This state class should not be used for sensors where the absolute value is interesting instead of the accumulated growth or decline, for example remaining battery capacity or CPU load; in such cases state class measurement should be used instead.

Sometimes it's really hard to tell, which one to use. Thanks for the hint :+1: 
I will change that.

chriss158 committed:
Use state class measurement for illumination sensors

simonszu commented:
I really hoped that this PR would make it to the 2021.11 release the last time i checked it. 
Another month to wait 😔

chriss158 committed:
Merge remote-tracking branch 'upstream/dev' into homematic-long-term-statistics

chriss158 committed:
Move sensor entity desc missing warning to setup_platform

chriss158 commented on homeassistant/components/homematic/entity.py:
```diff
@@ -27,20 +30,27 @@
         self._state = config.get(ATTR_PARAM)
         self._unique_id = config.get(ATTR_UNIQUE_ID)
-        self._data = {}
+        self._data: dict[str, str] = {}
```
I had to set a type for this, because mypy gave me an error.  I think this one should be a dict[str, str] or am I wrong?

chriss158 committed:
Set type for hmdevice and homematic to fix mypy error

pvizeli commented on homeassistant/components/homematic/entity.py:
```diff
@@ -27,20 +30,27 @@
+    def __init__(
+        self,
+        config: dict[str, str],
+        entity_description: SensorEntityDescription | None = None,
```
Use the base class from `from homeassistant.helpers.entity import Entity, EntityDescription` as type

pvizeli reviewed (approved):
Good to merge after the Type is fixed on the base class

chriss158 committed:
Use EntityDescription instead of SensorEntityDescription

chriss158 committed:
Merge remote-tracking branch 'upstream/dev' into homematic-long-term-statistics

pvizeli committed:
Update entity.py

pvizeli committed:
fix type

chriss158 commented:
Just curious. Why is the None type for `hmdevice` and `homeassistant` necessary?

pvizeli committed:
Update climate.py

pvizeli committed:
fix v2
